src/helpers/approval-utils.js

```diff
@@ -0,0 +1,58 @@
+import {BigNumber as BN} from "bignumber.js";
+
+export const isExchangeUnapproved = opts => {
+  const {
+    value,
+    currency,
+    drizzleCtx,
+    account,
+    contractStore,
+    exchangeAddresses,
+  } = opts;
+  if (!currency || currency === 'ETH') {
+    return false;
+  }
+  const inputExchange = exchangeAddresses.fromToken[currency];
+  if (!inputExchange) {
+    return false;
+  }
+
+  const allowanceKey = drizzleCtx.contracts[inputExchange].methods.allowance.cacheCall(account, inputExchange);
+  const allowance = contractStore[inputExchange].allowance[allowanceKey];
+  if (!allowance) {
+    return false;
+  }
+
+  return BN(value).isGreaterThan(BN(allowance.value));
+};
+
+export const approveExchange = opts => {
+  const {
+    currency,
+    drizzleCtx,
+    account,
+    exchangeAddresses,
+  } = opts;
+  const inputExchange = exchangeAddresses.fromToken[currency];
+  console.log(inputExchange);
+  if (!inputExchange) {
+    return;
+  }
+
+  return drizzleCtx.contracts[inputExchange].methods.approve.cacheSend(
+    inputExchange,
+    300, // TODO: what is this number?
```

Comment:
This would be the `decimals * 10 ** 18` number

Reply:
`decimals * 10**8`

although `decimals` frequently equals `10**18`

Reply:
Just a quick note maybe we should rename `decimals` to `decimalsMultiplier` or something. Since its technically:

`input_decimals = input_tolen_contract.methods.decimals()`
`decimals = 10 ** input_decimals`

I'm starting to think calling it `decimals` is a bit confusing.

Reply:
Added decimals into number calc (didn't rename though) here: https://github.com/Uniswap/uniswap-frontend/pull/66/files#diff-e678addf88010a62326dbec0e20d6176R48